File: algorithms/dfs.py
from algorithms import algorithm

class DFS(algorithm.Algorithm):
    def __init__(self, graph, start_node, end_node):
        super().__init__(graph, start_node, end_node)
        self.stack = [self.start_node]
        self.visited = {self.start_node}
        self.came_from = {self.start_node: None}
    
    def solve(self):
        self.frontier.add(self.start_node)
        yield self.frontier, self.visited, self.start_node, [self.start_node]

        while self.stack:
            current_node = self.stack.pop()
            self.frontier.discard(current_node)
            self.visited.add(current_node)

            if current_node == self.end_node:
                yield self.frontier, self.visited, current_node, self.reconstruct_path()
                return
            
            for neighbor in range(len(self.graph[current_node])):
                if self.graph[current_node][neighbor] > 0 and neighbor not in self.visited and neighbor not in self.frontier:
                    self.frontier.add(neighbor)
                    self.came_from[neighbor] = current_node
                    self.stack.append(neighbor)
        
            yield self.frontier, self.visited, current_node, self.reconstruct_path(current_node)

    def reconstruct_path(self, target_node=None):
        if target_node is None:
            target_node = self.end_node
        path = []
        current_node = target_node
        while current_node is not None:
            path.append(current_node)
            current_node = self.came_from[current_node]
        path.reverse()
        return path

File: src/app/core/algorithms/dfs.ts
import { Algorithm, NodeState } from './algorithm';

export class DFS extends Algorithm {
  private stack: number[] = [];
  private visitedSet: Set<number> = new Set();
  private frontierSet: Set<number> = new Set();
  private cameFrom: Map<number, number | null> = new Map();

  constructor(graph: number[][], startNode: number, endNode: number) {
    super(graph, startNode, endNode);
    this.stack.push(this.startNode);
    this.visitedSet.add(this.startNode);
    this.cameFrom.set(this.startNode, null);
  }

  *solve(): Generator<NodeState, void, unknown> {
    this.frontierSet.add(this.startNode);
    yield {
      frontier: new Set(this.frontierSet),
      visited: new Set(this.visitedSet),
      currentNode: this.startNode,
      path: [this.startNode]
    };

    while (this.stack.length > 0) {
      const currentNode = this.stack.pop()!;
      this.frontierSet.delete(currentNode);
      this.visitedSet.add(currentNode);

      if (currentNode === this.endNode) {
        yield {
          frontier: new Set(this.frontierSet),
          visited: new Set(this.visitedSet),
          currentNode: currentNode,
          path: this.reconstructPath(currentNode)
        };
        return;
      }

      for (let neighbor = 0; neighbor < this.graph[currentNode].length; neighbor++) {
        const weight = this.graph[currentNode][neighbor];
        if (weight > 0 && !this.visitedSet.has(neighbor) && !this.frontierSet.has(neighbor)) {
          this.frontierSet.add(neighbor);
          this.cameFrom.set(neighbor, currentNode);
          this.stack.push(neighbor);
        }
      }

      yield {
        frontier: new Set(this.frontierSet),
        visited: new Set(this.visitedSet),
        currentNode: currentNode,
        path: this.reconstructPath(currentNode)
      };
    }
  }

  private reconstructPath(targetNode: number): number[] {
    const path: number[] = [];
    let current: number | null = targetNode;
    while (current !== null) {
      path.push(current);
      current = this.cameFrom.get(current) ?? null;
    }
    return path.reverse();
  }
}
